### ISSUE TITLE
No obvious way to dismiss drawer when using TalkBack

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#1715]

**Comment by goderbauer** [comment]
This was recently fixed for iOS. We probably have to do something similar on Android.

**Comment by goderbauer** [comment]
Here is what's happening: when the "dismiss" button (that's the area to the right of the drawer) is double tapped in accessibility mode, that tap event is not send to the dismiss button. It's instead send to the thing that's currently in the center of the dismiss button. Since the drawer is overlaying the dismiss button, the click event is send to whatever element from the drawer is currently at the center of the dismiss button.

I fixed a similar issue on iOS by specifically dispatching the tap to the right element. Looks like on Android it's already setup to do that: https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L156.
However, for reasons not clear to me, the surrounding `performAction` method is only called for focus changes. The double tap (or CLICK action) never makes its way to `performAction` and so far I haven't found a way to intercept it and dispatch it to the right element... :(

**Comment by goderbauer** [comment]
Android translates the double tap in accessibility mode to simulated touch events, which are not processed by `performAction`. Instead, they are processed by the view's `onTouchEvent` method.

This is what `onTouch` sees:
```
I/FlutterView(22190): touch event MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=0, x[0]=720.0, y[0]=1274.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=59718012, downTime=59718012, deviceId=7, source=0x1002 }
I/FlutterView(22190): touch event MotionEvent { action=ACTION_UP, actionButton=0, id[0]=0, x[0]=720.0, y[0]=1274.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=59718012, downTime=59718012, deviceId=7, source=0x1002 }
```

And `onTouch` then dispatches that touch event to whatever is topmost at the provided coordinates (which might not be the accessibility-selected widget).

To make matters even worse: The "simulated" events seem to be indistinguishable from real touch events (which even in accessibility mode can be generated by tapping the screen with two fingers). That means, in `onTouch` it's unclear if a touch event should be dispatched to the topmost widget at the given position (touching the screen somewhere with two fingers) or to the currently accessibility-selected element (simulated touch events from double tapping the screen).

**Comment by goderbauer** [comment]
After talking to the Android Accessibility Team:

Simple solution to this (and the related #1716 issue): Modal Barriers in Android should not be exposed as clickable to dismiss the modal. Instead, the android back button should be used to close the modal. See other Android apps with a drawer, e.g. Gmail.

Ugly solution: The issue we are facing is only present with custom views in Android N or earlier. Native views can detect that the simulated `MotionEvent`s send to the view's `onTouchEvent` are meant for the currently accessibility-focused view. All other views ignore these events. Unfortunately, the information if a `MotionEvent` is meant to be handled only by the currently accessibility-focused node is not public (we could get to it via reflection #ugly). 

Future solution: Starting with Android O, Android changes its behavior here in our favor: Instead of sending simulated touch events on double tap, a proper `ACTION_CLICK` is send to the `performAction` method of the currently selected accessibility node. With that change, our current implementation works as expected and the drawer is properly closed when the modal barrier is double-tapped.

Suggestion for moving forward: Let's not expose the modal barrier as clickable on Android in accessibility mode and instead rely on the back button to dismiss it. This is in line with how native apps deal with modal barriers as well and represents what the users expects on Android. If this becomes a problem with activating other virtual views that are partially covered by another view, we might have to resort to the ugly reflection trick (or wait for everybody to upgrade to Android O).
